tensorflow/compiler/xla/service/pattern_matcher_test.cc

```diff
@@ -1102,5 +1102,94 @@ TEST_F(PatternMatcherTest, CustomCallMatchers) {
       root, m::CustomCall("test_target", m::Parameter(1), m::Parameter(0))));
 }
 
+TEST_F(PatternMatcherTest, OptionalUnaryOp) {
+  constexpr char kModuleStr[] = R"(
+    HloModule test_module
+
+    ENTRY test {
+      p0 = f32[] parameter(0)
+      cos = cosine(p0)
+      ROOT out = f32[] abs(cos)
+    }
+  )";
+  TF_ASSERT_OK_AND_ASSIGN(auto hlo_module,
+                          ParseAndReturnVerifiedModule(kModuleStr));
+  auto* root = hlo_module->entry_computation()->root_instruction();
+
+  EXPECT_TRUE(Match(
+      root,
+      m::OptionalUnaryOp({HloOpcode::kBitcast, HloOpcode::kCos}, m::Abs())));
+  EXPECT_TRUE(
+      Match(root, m::OptionalUnaryOp({HloOpcode::kBitcast, HloOpcode::kCos},
+                                     m::Abs(m::Cos()))));
+  EXPECT_TRUE(Match(
+      root, m::OptionalUnaryOp({HloOpcode::kCos, HloOpcode::kAbs}, m::Cos())));
+  EXPECT_FALSE(Match(
+      root,
+      m::OptionalUnaryOp({HloOpcode::kCos, HloOpcode::kAbs}, m::Bitcast())));
+  EXPECT_FALSE(Match(
+      root,
+      m::OptionalUnaryOp({HloOpcode::kCos, HloOpcode::kBitcast}, m::Cos())));
+
+  std::string description = absl::StrCat(
+      "an HloInstruction which optionally matches a unary operand with one of "
+      "the opcodes {",
+      HloOpcodeString(HloOpcode::kCos), ", ", HloOpcodeString(HloOpcode::kAbs),
+      "} before matching the given pattern, an HloInstruction with opcode ",
+      HloOpcodeString(HloOpcode::kBitcast), ".");
+  std::string explanation = absl::StrCat(
+      "HloInstruction doesn't have opcode ",
+      HloOpcodeString(HloOpcode::kBitcast),
+      "\nin cos = f32[] cosine(f32[] p0) ",
+      "\nExpected the operand of the instruction to match the prescribed "
+      "operand pattern. HloInstruction doesn't have opcode ",
+      HloOpcodeString(HloOpcode::kBitcast),
+      "\nin out = f32[] abs(f32[] cos)\nin out = f32[] abs(f32[] cos)");
```

Comment:
This doesn't seem like a good error?

 - we repeat "in" twice, but that doesn't make sense.
 - there's no conjunction connecting between "expected it to match the prescribed pattern" and the description of what the pattern is.
 - same as above, saying "the prescribed pattern" is unnecessary, we can just say what the pattern is.